packages/sui-gateway/src/gateway/mod.rs

```diff
@@ -0,0 +1,178 @@
+    fn from(bytes: [u8; 32]) -> Self {
+        Self { bytes }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct WeightedSigner {
+    pub pub_key: Vec<u8>,
+    pub weight: u128,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WeightedSigners {
+    pub signers: Vec<WeightedSigner>,
+    pub threshold: u128,
+    pub nonce: Bytes32,
+}
+
+impl TryFrom<VerifierSet> for WeightedSigners {
+    type Error = Report<Error>;
+
+    fn try_from(verifier_set: VerifierSet) -> Result<Self, Self::Error> {
+        let mut signers = verifier_set
+            .signers
+            .values()
+            .map(|signer| match &signer.pub_key {
+                PublicKey::Ecdsa(key) => Ok(WeightedSigner {
+                    pub_key: key.to_vec(),
+                    weight: signer.weight.into(),
+                }),
+                PublicKey::Ed25519(_) => Err(Report::new(Error::UnsupportedPublicKey)),
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        signers.sort_by(|signer1, signer2| signer1.pub_key.cmp(&signer2.pub_key));
+
+        let nonce = Uint256::from(verifier_set.created_at).to_be_bytes().into();
+
+        Ok(Self {
+            signers,
+            threshold: verifier_set.threshold.into(),
+            nonce,
+        })
+    }
+}
+
+impl WeightedSigners {
+    pub fn hash(&self) -> [u8; 32] {
+        let hash =
+            Keccak256::digest(bcs::to_bytes(&self).expect("failed to serialize WeightedSigners"));
+
+        hash.into()
+    }
+}
+
+#[derive(Serialize)]
+pub struct MessageToSign {
+    pub domain_separator: Bytes32,
+    pub signers_hash: Bytes32,
+    pub data_hash: Bytes32,
+}
+
+impl MessageToSign {
+    pub fn hash(&self) -> [u8; 32] {
+        let hash =
+            Keccak256::digest(bcs::to_bytes(&self).expect("failed to serialize MessageToSign"));
+
+        hash.into()
+    }
+}
+
+#[derive(Serialize)]
+pub struct Message {
+    source_chain: String,
+    message_id: String,
+    source_address: String,
+    destination_id: SuiAddress,
+    payload_hash: Bytes32,
+}
+
+impl TryFrom<&router_api::Message> for Message {
+    type Error = Report<Error>;
+
+    fn try_from(value: &router_api::Message) -> Result<Self, Self::Error> {
+        Ok(Self {
+            source_chain: value.cc_id.source_chain.to_string(),
+            message_id: value.cc_id.message_id.to_string(),
+            source_address: value.source_address.to_string(),
+            destination_id: value.destination_address.parse()?,
+            payload_hash: value.payload_hash.into(),
+        })
+    }
+}
+
+#[derive(Serialize)]
+pub struct Signature {
+    bytes: Vec<u8>,
+}
+
+impl TryFrom<multisig::key::Signature> for Signature {
+    type Error = Report<Error>;
+
+    fn try_from(signature: multisig::key::Signature) -> Result<Self, Self::Error> {
+        match signature {
+            multisig::key::Signature::EcdsaRecoverable(signature) => Ok(Self {
```

Comment:
Add a comment that this is only being called during encoding after signatures have been normalized to EcdsaRecoverable

Reply:
It's not clear when just looking at this implementation this invariant holds. Maybe it should support converting from both

Reply:
Let me see how hard it is to make it work for both. Kinda like that idea.

Reply:
Ok don't think handling both is realistic here. It requires the pub key and msg to convert from non-recoverable to recoverable. I'll add the comment.